Mathlib/Order/Birkhoff.lean

```diff
@@ -0,0 +1,228 @@
+powerset lattice. -/
+noncomputable def birkhoffSet : α ↪o Set {a : α // SupIrred a} := by
+  by_cases IsEmpty α
+  · exact ⟨⟨isEmptyElim, isEmptyElim⟩, isEmptyElim ‹α›, isEmptyElim ‹α›⟩
+  rw [not_isEmpty_iff] at h
+  have := Fintype.toOrderBot α
+  exact OrderIso.lowerSetSupIrred.toOrderEmbedding.trans ⟨⟨_, SetLike.coe_injective⟩, Iff.rfl⟩
+
+/-- **Birkhoff's Representation Theorem**. Any finite distributive lattice can be embedded in a
+powerset lattice. -/
+noncomputable def birkhoffFinset : α ↪o Finset {a : α // SupIrred a} := by
+  exact (birkhoffSet _).trans Fintype.finsetOrderIsoSet.symm.toOrderEmbedding
+
+variable {α}
+
+@[simp] lemma coe_birkhoffFinset (a : α) : birkhoffFinset α a = birkhoffSet α a := by
+  classical
+  simp [birkhoffFinset]
+  rw [OrderIso.coe_toOrderEmbedding, Fintype.coe_finsetOrderIsoSet_symm]
+  simp
+
+@[simp] lemma birkhoffSet_sup (a b : α) :
+    birkhoffSet α (a ⊔ b) = birkhoffSet α a ∪ birkhoffSet α b := by
+  unfold OrderEmbedding.birkhoffSet; split <;> simp
+
+@[simp] lemma birkhoffSet_inf (a b : α) :
+    birkhoffSet α (a ⊓ b) = birkhoffSet α a ∩ birkhoffSet α b := by
+  unfold OrderEmbedding.birkhoffSet; split <;> simp
+
+variable [DecidableEq α]
+
+@[simp] lemma birkhoffFinset_sup (a b : α) :
+    birkhoffFinset α (a ⊔ b) = birkhoffFinset α a ∪ birkhoffFinset α b := by
+  dsimp [OrderEmbedding.birkhoffFinset]
+  rw [birkhoffSet_sup, OrderIso.coe_toOrderEmbedding]
+  simpa using OrderIso.map_sup _ _ _
+
+@[simp] lemma birkhoffFinset_inf (a b : α) :
+    birkhoffFinset α (a ⊓ b) = birkhoffFinset α a ∩ birkhoffFinset α b := by
+  dsimp [OrderEmbedding.birkhoffFinset]
+  rw [birkhoffSet_inf, OrderIso.coe_toOrderEmbedding]
+  simpa using OrderIso.map_inf _ _ _
+
+variable [OrderBot α]
+
+@[simp] lemma birkhoffSet_apply (a : α) : birkhoffSet α a = OrderIso.lowerSetSupIrred a := by
+  simp [birkhoffSet]; convert rfl
+
+end OrderEmbedding
+
+namespace LatticeHom
+variable [Fintype α] [DecidableEq α] [@DecidablePred α SupIrred]
+
+/-- **Birkhoff's Representation Theorem**. Any finite distributive lattice can be embedded in a
+powerset lattice. -/
+noncomputable def birkhoffSet : LatticeHom α (Set {a : α // SupIrred a}) where
+  toFun := OrderEmbedding.birkhoffSet α
+  map_sup' := OrderEmbedding.birkhoffSet_sup
+  map_inf' := OrderEmbedding.birkhoffSet_inf
+
+/-- **Birkhoff's Representation Theorem**. Any finite distributive lattice can be embedded in a
+powerset lattice. -/
+noncomputable def birkhoffFinset : LatticeHom α (Finset {a : α // SupIrred a}) where
+  toFun := OrderEmbedding.birkhoffFinset α
+  map_sup' := OrderEmbedding.birkhoffFinset_sup
+  map_inf' := OrderEmbedding.birkhoffFinset_inf
+
+lemma birkhoffFinset_injective : Injective (birkhoffFinset α) :=
+  (OrderEmbedding.birkhoffFinset α).injective
+
+end LatticeHom
+
+lemma exists_birkhoff_representation.{u} (α : Type u) [Finite α] [DistribLattice α] :
+    ∃ (β : Type u) (_ : DecidableEq β) (_ : Fintype β) (f : LatticeHom α (Finset β)),
```

Comment:
This `DecidableEq` is not true constructively, so I'd rather leave it off.
```suggestion
    ∃ (β : Type u) (_ : Fintype β) (f : LatticeHom α (Finset β)),
```

Reply:
This is needed to state `f : LatticeHom α (Finset β)`.